design/0001-frr.md

```diff
@@ -0,0 +1,411 @@
+│               │  │                    │               │  └─┤  <<package>>  │
+│     nodes     │  │                    │   Protocol    │    │               │
+│    <<Node>>   ├──┘                    │ <<interface>> │    └───────────────┘
+│               │                       │               │
+└───────────────┘                       └───────▲───────┘
+                                                │<<implements>>
+                                     ┌──────────┴─────────┐
+                                     │                    │
+                              ┌──────┴────────┐  ┌────────┴──────┐
+                              │     main::    │  │    main::     │
+                              │ L2_controller │  │ bgp_controller│
+                              │   <<class>>   │  │   <<class>>   │
+                              │               │  │               │
+                              └───────────────┘  └───────┬───────┘
+                                                         │
+                                                         │  <<use>>
+                                                 ┌───────▼───────┐
+                                                 │               │
+                                                 │    session    │
+                                                 │ <<interface>> │
+                                                 │               │
+                                                 └───────▲───────┘
+                                                         │ <<implement>>
+                                             ┌───────────┴─────────────┐
+                                             │                         │
+                                     ┌───────┴───────┐         ┌───────┴───────┐
+                                     │               │         │               │
+                                     │  bgp_metallb  │         │    bgp_frr    │
+                                     │  <<package>>  │         │  <<package>>  │
+                                     │               │         │               │
+                                     └───────────────┘         └───────┬───────┘
+                                                                       │
+                                                                       │
+                                                               ┌───────┴───────┐
+                                                               │               │
+                                                               │      frr      │
+                                                               │ <<container>> │
+                                                               │               │
+                                                               └───────────────┘
+
+```
+
+#### FRR deployment
+
+FRR will be deployed as a container. The current proposal is to deploy it
```

Comment:
I had been thinking about deploying it as a separate pod as it will 1) make it easier to independently manage the lifecycle of MetalLB and 2) make it easier for a user to use their own MetalLB pod. However, I don't really have any strong convictions about this at the moment.

Regarding communication, `vtysh` (the FRR command line tool) communicates with the daemons using a unix socket (I need to confirm if there is a TCP socket option). The gRPC interface is over a TCP socket.

Reply:
Can you describe the security of the vtysh interface if they're running in separate pods?  A benefit of sticking them in the same pod is a simpler security model for the communication.

Reply:
> A nice side effect of running in a separate pod would be that the metallb speaker would no longer require host network access. Only the FRR pod would be on the host network.

Yes, good point. I will add this point in the next revision.

Reply:
> Can you describe the security of the vtysh interface if they're running in separate pods? A benefit of sticking them in the same pod is a simpler security model for the communication.

Another good point. I did some investigation (and I am not an expert). vtysh is available over unix sockets or tcp sockets. 

For unix sockets, the socket is protected by unix file permissions.

For tcp sockets, it is protected by a password that is specified in the config file. It should be noted, on further investigation, that for tcp sockets you need to open a telnet session to a particular port and that is now the remote connectivity can be achieved. There is no native support for something like SSL that I can see. You can also limit what IP address that listens for connections (e.g. 127.0.0.1 or 0.0.0.0, etc)

Reply:
I really like the separate lifecycle management benefit in particular.  I've also seen requirements for "TLS everywhere" before though, so this could come back to bite us.  Even if it's localhost only, it still doesn't seem secure enough.

My feeling right now is to stick it within the same pod to keep an acceptable security model.  Hopefully we can do something better in the future with the grpc interface?  Unfortunately from a quick look, the grpc interface doesn't support anything better.  We may have an easier time sticking a secure proxy in front of it with grpc.  In any case, if we need to improve this in frr, putting that energy on the grpc side is probably a better investment.

Reply:
Yes, I agree that this is something to flag to the FRR community.

On the community call this week, we discussed this. The general consensus was that we should use the TCP/IP interface on 127.0.0.1 and deploy as a container in an MetalLB pod. This reduces the complexity from a Metal LB project perspective. If a consumer wants to deploy FRR as a pod they could then (somehow) change the IP address used to connect to FRR.

Reply:
Is it possible to use a socket file ?

Reply:
Socket file is another supported communication method.

Reply:
I think this settled on running it within the speaker pod, right?

Reply:
I think so. I will summarize this conversation and commit an update.